File: framework/core/src/main/java/cn/wjybxx/bigcat/pb/MessageCodec.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.bigcat.pb;

import cn.wjybxx.base.ObjectUtils;
import cn.wjybxx.dson.text.ObjectStyle;
import cn.wjybxx.dsoncodec.DsonCodec;
import cn.wjybxx.dsoncodec.DsonObjectReader;
import cn.wjybxx.dsoncodec.DsonObjectWriter;
import cn.wjybxx.dsoncodec.TypeInfo;
import cn.wjybxx.dsoncodec.annotations.DsonCodecScanIgnore;
import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.MessageLite;
import com.google.protobuf.Parser;

import javax.annotation.Nonnull;
import java.util.Objects;
import java.util.function.Supplier;

/**
 * Message会序列化为字节数组，因此不可以作为顶层对象。
 *
 * @author wjybxx
 * date 2023/4/2
 */
@DsonCodecScanIgnore
public class MessageCodec<T extends MessageLite> implements DsonCodec<T> {

    private final Class<T> clazz;
    private final Parser<T> parser;

    public MessageCodec(Class<T> clazz, Parser<T> parser) {
        this.clazz = clazz;
        this.parser = Objects.requireNonNull(parser, "parser");
    }

    @Nonnull
    @Override
    public Class<T> getEncoderClass() {
        return clazz;
    }

    @Override
    public void writeObject(DsonObjectWriter writer, T instance, TypeInfo<?> typeArgInfo, ObjectStyle style) {
        writer.writeBytes(null, instance.toByteArray());
    }

    @Override
    public T readObject(DsonObjectReader reader, TypeInfo<?> typeArgInfo, Supplier<? extends T> factory) {
        byte[] bytes = reader.readBytes(reader.getCurrentName());
        if (bytes == null) {
            return null;
        }
        try {
            return parser.parseFrom(bytes);
        } catch (InvalidProtocolBufferException e) {
            return ObjectUtils.rethrow(e);
        }
    }

}

File: csharp/Wjybxx.BigCat.Core/src/Util/MessageCodec.cs
#region LICENSE

// Copyright 2025 wjybxx(845740757@qq.com)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;
using Google.Protobuf;
using Wjybxx.Dson.Codec;
using Wjybxx.Dson.Text;

namespace Wjybxx.BigCat.Util
{
/// <summary>
/// protobuf的Message的解码器
/// Message会序列化为字节数组，因此不可以作为顶层对象。
/// (Message作为Rpc方法的参数或结果时会被特殊处理，不会走用户的序列化。)
/// 
/// C#的自循环泛型，其如何运行的更难理解。
/// </summary>
/// <typeparam name="T"></typeparam>
public class MessageCodec<T> : IDsonCodec<T> where T : IMessage<T>
{
    private readonly MessageParser<T> parser;

    public MessageCodec(MessageParser<T> parser) {
        this.parser = parser;
    }

    public Type GetEncoderType() => typeof(T);

    public bool AutoStartEnd => false;

    public void WriteObject(IDsonObjectWriter writer, in T inst, Type declaredType, ObjectStyle style) {
        writer.WriteBytes(null, inst.ToByteArray());
    }

    public T ReadObject(IDsonObjectReader reader, Func<object>? factory = null) {
        byte[] bytes = reader.ReadBytes(reader.CurrentName);
        if (bytes == null) {
            return default;
        }
        return parser.ParseFrom(bytes);
    }
}
}